### DISCUSSION TITLE
I will be using AceWire in SSD1306Ascii

### DISCUSSION THREAD

**Discussion by greiman** [bxparks/AceWire#12] [General]
This isn't an issue.  I just wanted to let you know how a template was developed for SSD1306AsciiWire that almost exactly matches yours.

I have had several users that want to use SSD1306Ascii on pins other than SCL, SDA.  One submitted a pull request for [SoftWire](https://github.com/stevemarple/SoftWire).  I decided to make the base class for using Wire a template, SSD1306AsciiWireBase, so other libraries could be used.

I then started looking for software and hardware libraries I could use in examples.  I found AceWire.  I then discovered your fork of SSD1306Ascii.

Your SSD1306AsciiAceWire.h is almost identical to my SSD1306AsciiWireBase.h.  I use WireType you use T_WIREI for the type.

I may also do the same for SPI.  Fewer people use SPI displays but I have a SoftSPI so it's likely I will make a SPI template so it is easy to use AceSPI and other SPI libraries.

**Comment by bxparks** [comment]
@greiman : (Sorry for the delay. You posted this while I was on vacation, and it got progressively buried in my inbox, and I was busy with other things when I got back.) Glad that you found AceWire and AceSPI useful. 

Just an FYI, I'm relatively sure that the interface for AceWire is stable. But AceSPI currently supports only transmit, not receive. I'm planning to fix that in near future (next few months?), and I *think* I can make everything backwards compatible by converting the current `void XxxSpiInterface::transfer()` and `void XxxSpiInterface::transfer16()` methods to return an `uint8_t` and `uint16_t`, instead of `void`.

**Reply by greiman** [reply]
I have been upgrading a software SPI and software I2C I did many years ago for [this](https://github.com/greiman/DigitalIO) library.

I am upgrading the AVR fast digital.  I implement it with constexpr arrays and functions for the pin map.  Here is digitalWriteFast which is interrupt safe and the constexpr pin map functions.  For most AVR ports with constant arguments this compiles into two bytes.

```
void badPinNumber()
    __attribute__((error("Pin number is too large or not a constant")));

[[gnu::always_inline]] inline void badPinCheck(uint8_t pin) {
  if (!__builtin_constant_p(pin) || pin >= GPIO_PIN_MAP_SIZE) {
    badPinNumber();
  }
}

constexpr uint8_t pinMask(uint8_t pin) { return GpioPinMap[pin].mask; }

constexpr volatile uint8_t* portReg(uint8_t pin) {
  return GpioPinMap[pin].port;
}

// Note the Arduino API with uint8_t val.
[[gnu::always_inline]] inline void digitalWriteFast(uint8_t pin, uint8_t val) {
  badPinCheck(pin);
  if (int(portReg(pin)) < 0X40) {
    // uses the interrupt safe sbi or cbi instructions.
    val ? *portReg(pin) |= pinMask(pin) : *portReg(pin) &= ~pinMask(pin);
    return;
  }
  ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
   // higher port require interrupt protection.
    val ? *portReg(pin) |= pinMask(pin) : *portReg(pin) &= ~pinMask(pin);
  }
}
// similar functions for ddrWriteFast(), digitalReadFast(), and pinModeFast().
```

I have an example program that can run on any AVR board to generate the GpioPinMap array.  I include a map for common boards.

My old I2C API was very basic I2C, not the Wire API.  I am implementing a Wire API as a template where the open drain driver is a class argument.  I have implemented several open drain driver classes.

I plan to do the same for my software SPI.

**Comment by greiman** [comment]
Here is an example of the [DigitalO](https://github.com/greiman/DigitalIO) SoftSPI.  It is on an Uno sending a 0X55 byte in mode zero .

![SoftSPI](https://user-images.githubusercontent.com/3964207/236688353-448cb27f-2233-4288-994a-03658b77380f.jpg)

I am updating it to use more modern C++.

**Reply by bxparks** [reply]
I have been frustrated many times with the poor performance of `digitalWrite()` and `digitalRead()` functions on AVR processors. It makes it difficult to write bigbanging routines on general GPIO pins.

The problem I have seen with the various `digitalWriteFast()` libraries is that there are multiple versions floating around the internet, and it is not obviously clear which one is best, and which ones are being maintained. There are now at least 3 major ones that I am aware of:

* https://github.com/watterott/Arduino-Libs/tree/master/digitalWriteFast
    * this version might be close to the original version
    * not uploaded to the Arduino Library Manager, so more work to use
    * probably not maintained anymore
* https://github.com/NicksonYap/digitalWriteFast
   * this fork was the one in the Arduino Library Manager, but
   * completely abandoned and unmaintained as far as I can tell, last updated 8 years ago
* https://github.com/ArminJo/digitalWriteFast
   * a new fork of watterott's version?
   * this seems to be the version that is now served by the Arduino Library Manager

 When I was researching this a few years ago as I was learning Arduino, I came across your DigitalIO library, but I did not consider using it because:
* the API seemed too different from the standard Arduino API: `digitalWrite()`, `digitalRead()`, `pinMode()` etc.
* the library seemed to pull in more functionality than I wanted (pulling in I2C and SPI, which I did not understand as a beginner)

I have ~~had~~ thought about just diving in and writing my own version of `digitalWriteFast()` and `digitalReadFast()`, but I have always held back because I was not sure that I could do a better job than many other people who have tried before me. And I was not sure that I could test and maintain such a library across so many different AVR boards. I had to ask myself, "Is this the best use of my time? There are so many other interesting projects to do..."

Your code snippet suggests that you are extracting the various pin-to-port mapping arrays that are defined in the `pins_arduino.h` file of each board, then copying these values into your library's `GpioPinMap`. I have 2 passing questions:

1) Do you think it would be possible to programmatically extract those maps from the C++ source code, and then auto-generate the necessary data structures and files needed by this new `digitalWriteFast` library? 

   Maybe a Python script that parses the `pins_arduino.h` of each board, and then generates the equivalent of your `GpioPinMap`. Seems like the more we can automate this process, the fewer bugs there will be, and the easier it would be maintain the library.

   Another random idea, my [EpoxyDuino](https://github.com/bxparks/EpoxyDuino) tool allows Arduino programs to run on Unix-like machines. Maybe I could process the various `pins_arduino.h` under EpoxyDuino and programmatically generate the pin-mapping data structures.

2) I don't have full understanding of the AVR macros related to interrupt safety. But it seems like the `ATOMIC_BLOCK(ATOMIC_RESTORESTATE)` macro would generate extra machine code for each call to `digitalWriteFast()`. My question is, is it absolutely necessary for the various `digitalWriteFast()`, `digitalReadFast()`, `pinModeFast()` functions to be interrupt safe? Is it possible to push this problem to the calling client code, where the calling code can choose to turn off interrupts before an SPI or I2C transmission, then turn it on again after it is done with the `digitalWriteFast()` functions?

**Comment by greiman** [comment]
> Do you think it would be possible to programmatically extract those maps from the C++ source code, and then auto-generate the necessary data structures and files needed by this new digitalWriteFast library?

I don't think extracting from C++ code is practical.  I wrote a sketch that generates pin maps by running on a given board.   

> I don't have full understanding of the AVR macros related to interrupt safety. But it seems like the ATOMIC_BLOCK(ATOMIC_RESTORESTATE) macro would generate extra machine code for each call to digitalWriteFast(). My question is, is it absolutely necessary for the various digitalWriteFast(), digitalReadFast(), pinModeFast() functions to be interrupt safe? 

Unfortunately many libraries respond to interrupts using digitalWriteFast.  If an interrupt occurs during a read update write by a background program and the pins for the background and interrupt routine are in the same 8-bit PORT register an undefined result will occur.  

Most PORT registers  on AVR boards are at low addresses and AVR has two interrupt safe instructions sbi/cbi to set or clear a bit with one atomic instruction.  The Mega is an exception.

There is not a clean way to solve this problem since there are versions of digitalWriteFast that are interrupt save and others that are not.

I have used the special two cycle sib/cbi instructions where possible and ATOMIC_BLOCK where it is not safe.  The high registers already are much slower since they require many instructions to fetch update and store the result.

ATOMIC_BLOCK(ATOMIC_RESTORESTATE) adds the following.  It uses assembler so I will use the equivalent C++.

At the beginning of the block:
```
  uint8_t __s = SREG;
  cli();  // one byte disable interrupts
```
At the end of the block:
`SREG = __s;`

Here is the program that generates the pin map for a board:
```
#if defined(__AVR_ATmega4809__)
//  defined(ARDUINO_AVR_UNO_WIFI_REV2) || defined(ARDUINO_AVR_NANO_EVERY)
void boardPinMap() {
  // See file pins_arduino.h for above variants
  const uint8_t PIN_MAP_SIZE = NUM_DIGITAL_PINS + NUM_I2C_PINS + 2;
  for (uint8_t pin = 0; pin < PIN_MAP_SIZE; pin++) {
    char P = 'A' + digitalPinToPort(pin);
    Serial.print("    {&VPORT");
    Serial.print(P);
    Serial.print("_IN, &VPORT");
    Serial.print(P);
     Serial.print("_DIR, &VPORT");
    Serial.print(P);
    Serial.print("_OUT, 1 << ");
    uint8_t b = digitalPinToBitPosition(pin);
    Serial.print(b);
    if (pin < (PIN_MAP_SIZE - 1)) {
      Serial.print("},  // D");
    } else {
      Serial.print("}   // D");
    }
    Serial.println(pin);
  }     
}
#else // Older AVR boards
void boardPinMap() {
  for (uint8_t pin = 0; pin < NUM_DIGITAL_PINS; pin++) {
    char P = 'A' - 1 + digitalPinToPort(pin);
    Serial.print("    {&PIN");
    Serial.print(P);
    Serial.print(", &DDR");
    Serial.print(P);
     Serial.print(", &PORT");
    Serial.print(P);
    Serial.print(", 1 << ");
    int m = digitalPinToBitMask(pin);
    int b = 0;
    for (b = 0; (1 << b) != m; b++) {}
     Serial.print(b);
    if (pin < (NUM_DIGITAL_PINS - 1)) {
      Serial.print("},  // D");
    } else {
      Serial.print("}   // D");
    }
    Serial.println(pin);
  }
}
#endif  // boardPinMap()
void setup() {
  Serial.begin(9600);
  while(!Serial) {}
  Serial.println(
    "#ifndef BoardGpioPinMap_h\r\n"
    "#define BoardGpioPinMap_h\r\n"
    "static const GpioPinMap_t GpioPinMap[] = {");
  boardPinMap();
  Serial.println(
    "};\r\n"
    "#endif  // BoardGpioPinMap_h");  
}

void loop() {}
```

**Comment by greiman** [comment]
My DigitalIO library was started in 2008 before I knew about other libraries.  I did it to provide Software SPI for SdFat and various projects I was doing.

Here is a demo of a digitalWriteFast with the more common API.  It is a bit rough yet and is not a library.  Unzip the program and try it.  It generates a 125 ns wide pulse on an Uno and use about 12 bytes.

You can use the above program to produce more pin maps.  Click on the link to download it.

[DigitalWriteFastDemo.zip](https://github.com/bxparks/AceWire/files/11453052/DigitalWriteFastDemo.zip)
